include/openmc/mesh.h

```diff
@@ -399,17 +467,77 @@ class UnstructuredMesh : public Mesh {
   std::pair<moab::Tag, moab::Tag>
   get_score_tags(std::string score) const;
 
-  // data members
+  // Data members
   moab::Range ehs_; //!< Range of tetrahedra EntityHandle's in the mesh
   moab::EntityHandle tetset_; //!< EntitySet containing all tetrahedra
   moab::EntityHandle kdtree_root_; //!< Root of the MOAB KDTree
   std::unique_ptr<moab::Interface> mbi_; //!< MOAB instance
   std::unique_ptr<moab::AdaptiveKDTree> kdtree_; //!< MOAB KDTree instance
   std::vector<moab::Matrix3> baryc_data_; //!< Barycentric data for tetrahedra
+  std::vector<std::string> tag_names_; //!< Names of score tags added to the mesh
 };
 
 #endif
 
+#ifdef LIBMESH
+
+class LibMesh : public UnstructuredMesh {
+public:
+  // Constructors
+  LibMesh(pugi::xml_node node);
+  LibMesh(const std::string& filename);
+
+  // Methods
+  void bins_crossed(const Particle& p,
+                    std::vector<int>& bins,
+                    std::vector<double>& lengths) const override;
+
+  int get_bin(Position r) const override;
+
+  int n_bins() const override;
+
+  int n_surface_bins() const override;
+
+  std::pair<std::vector<double>, std::vector<double>>
+  plot(Position plot_ll, Position plot_ur) const override;
+
+  void add_score(const std::string& var_name) override;
+
+  void remove_scores() override;
+
+  void set_score_data(const std::string& var_name,
+                      const std::vector<double>& values,
+                      const std::vector<double>& std_dev) override;
+
+  void write(const std::string& base_filename) const override;
+
+  Position centroid(int bin) const override;
+
+  double volume(int bin) const override;
+
+  std::string library() const override;
+
+private:
+
+  void initialize() override;
+
+  //! Translate a bin value to an element reference
+  const libMesh::Elem& get_element_from_bin(int bin) const;
+
+  //! Translate an element pointer to a bin index
+  int get_bin_from_element(const libMesh::Elem* elem) const;
+
+  // Data members
+  std::unique_ptr<libMesh::Mesh> m_; //!< pointer to the libMesh mesh instance
+  std::vector<std::unique_ptr<libMesh::PointLocatorBase>> pl_; //!< per-thread point locators
+  std::unique_ptr<libMesh::EquationSystems> equation_systems_; //!< pointer to the equation systems of the mesh
+  std::string eq_system_name_; //!< name of the equation system holding OpenMC results
+  std::unordered_map<std::string, unsigned int> variable_map_; //!< mapping of variable names (tally scores) to libMesh variable numbers
+  libMesh::BoundingBox bbox_; //!< bounding box of the mesh
+  libMesh::Elem* first_element_; //!< pointer to the first element in the mesh (maybe should be a key?)
```

Comment:
Does the "maybe should be a key?" note indicate that anything remains to be addressed here? If not, I'd perhaps omit this or change the wording to explain what that means

Reply:
Thanks for pointing this out! There was still some work to be done here. In the end, what I wanted was the element ID according to libMesh.

Seeing as all I do with this pointer is get the element's ID later in `LibMesh::get_bin_from_element`, I'm just going to store that `dof_id_type` rather than the raw pointer.